### PR TITLE
feat: add mlx-audio integration — STT, TTS, STS support

### PR DESCRIPTION
**Pull request by ethannortharc**
> **Note to maintainer:** I noticed audio support (mlx-audio integration) was on the roadmap for oMLX. I had an immediate need for this functionality in my project, so I went ahead and implemented it on my fork. This PR offers that work back upstream in case it's useful. The implementation is functional but intentionally minimal — no performance optimizations have been done — so it should be straightforward to build on or adjust to fit your vision for the project. Happy to iterate on any feedback.

## Summary

oMLX currently supports LLM, VLM, Embedding, and Reranker model types. This PR adds full audio inference support by integrating [mlx-audio](https://github.com/Blaizzy/mlx-audio), enabling three new model categories:

- **STT (Speech-to-Text)** — Whisper, Qwen3-ASR, Parakeet, Voxtral
- **TTS (Text-to-Speech)** — Qwen3-TTS, Kokoro, F5-TTS, Sesame CSM, Dia, Spark, CosyVoice
- **STS (Speech-to-Speech)** — DeepFilterNet, MossFormer2, SAMAudio, LFM2.5-Audio

Audio models are auto-detected during model discovery, loaded through the existing EnginePool (with LRU eviction and memory management), and served via three new OpenAI-compatible API endpoints. `mlx-audio` is an optional dependency — existing functionality is completely unaffected when it is not installed.

This is a functional integration focused on correctness. No performance optimizations (batching, streaming, quantization-aware routing, etc.) have been attempted yet.

### New files (7 source + 7 test files)

| File | Lines | Purpose |
|------|-------|---------|
| `omlx/engine/stt.py` | 196 | STT engine — wraps `mlx_audio.stt.utils.load_model()` |
| `omlx/engine/tts.py` | 199 | TTS engine — wraps `mlx_audio.tts.utils.load_model()` |
| `omlx/engine/sts.py` | 366 | STS engine — family-specific adapters for 4 model types |
| `omlx/api/audio_routes.py` | 193 | Three FastAPI endpoints |
| `omlx/api/audio_models.py` | 54 | Pydantic request/response schemas |
| `tests/test_audio_stt.py` | 288 | STT endpoint + engine tests |
| `tests/test_audio_tts.py` | 261 | TTS endpoint + engine tests |
| `tests/test_audio_sts.py` | 469 | STS endpoint + engine + family detection tests |
| `tests/test_audio_discovery.py` | 452 | Model type detection tests (STT/TTS/STS) |
| `tests/test_audio_api.py` | 211 | `/v1/models` listing with audio models |
| `tests/test_audio_memory.py` | 312 | Memory management + LRU eviction for audio |
| `tests/fixtures/generate_test_audio.py` | 35 | Test audio generation utility |
| `tests/fixtures/test_audio.wav` | — | Test audio fixture |

### Modified files (8 files)

| File | Change |
|------|--------|
| `omlx/engine/__init__.py` | Export `STTEngine`, `TTSEngine`, `STSEngine` |
| `omlx/engine_pool.py` | Add `audio_stt`/`audio_tts`/`audio_sts` engine types; audio models get 0 KV headroom; engine instantiation routing |
| `omlx/model_discovery.py` | New `ModelType`/`EngineType` literals; `AUDIO_STT_*`, `AUDIO_TTS_*`, `AUDIO_STS_*` detection sets; detection logic in `detect_model_type()`; audio models no longer treated as "unsupported" |
| `omlx/server.py` | Register `audio_router` |
| `omlx/admin/routes.py` | Accept audio types in model settings; add mlx-audio to dashboard Engine Versions |
| `omlx/admin/templates/chat.html` | Filter audio/embedding/reranker from chat model selector |
| `omlx/admin/templates/dashboard/_modal_model_settings.html` | Audio type options in model settings dropdown |
| `pyproject.toml` | Add `audio` optional dependency group (`mlx-audio>=0.4.0`) |

Total: **+2,603 lines / -35 lines** across 21 files.

### How audio models integrate into oMLX

```
[Model Directory]
       |
   detect_model_type()          # config.json → "audio_stt" / "audio_tts" / "audio_sts"
       |
   EnginePool._register()       # engine_type = "audio_stt" / "audio_tts" / "audio_sts"
       |
   EnginePool._load_engine()    # → STTEngine / TTSEngine / STSEngine
       |
   Audio API Routes             # /v1/audio/transcriptions, /speech, /process
```

All three audio engines inherit from `BaseNonStreamingEngine` (not `BaseEngine`) because audio inference is a single forward pass — no KV cache, no streaming token generation. This means:

- **Memory**: 0 KV headroom allocated (vs. 25% for LLM/VLM)
- **Lifecycle**: Same LRU eviction, lazy loading, and `start()`/`stop()` pattern as other engines
- **Concurrency**: Model loading runs on the global MLX executor to avoid Metal command buffer races

### Engine implementations

**STTEngine** — Uses `mlx_audio.stt.utils.load_model()` (unified API). Input: audio file path. Output: `dict` with `text`, `language`, `duration`, `segments`.

**TTSEngine** — Uses `mlx_audio.tts.utils.load_model()` (unified API). Input: text + optional voice/speed. Output: WAV bytes. Handles both generator and direct array returns from different TTS model families.

**STSEngine** — No unified API exists in `mlx_audio.sts` (unlike STT/TTS). Each model family has a different class with a different processing method. The engine uses a family-specific adapter pattern:

| Family | Model Class | Load | Process | Output |
|--------|-------------|------|---------|--------|
| DeepFilterNet | `DeepFilterNetModel` | `.from_pretrained()` | `.enhance_file(in, out)` | WAV file |
| MossFormer2 | `MossFormer2SEModel` | `.from_pretrained()` | `.enhance(path)` | numpy array |
| SAMAudio | `SAMAudio` | `.from_pretrained()` | `.separate(audios, descriptions)` | SeparationResult |
| LFM2.5-Audio | `LFM2AudioModel` + `LFM2AudioProcessor` | `.from_pretrained()` ×2 | ChatState → generate → decode | Generator → WAV |

Family is auto-detected from the model name (e.g., "deepfilter" → deepfilternet, "mossformer" → mossformer2, "sam"+"audio" → sam_audio, "lfm" → lfm2).

### API endpoints

| Endpoint | Method | Input | Output | Compatibility |
|----------|--------|-------|--------|---------------|
| `/v1/audio/transcriptions` | POST | Multipart audio file + model | JSON (text, language, segments, duration) | OpenAI-compatible |
| `/v1/audio/speech` | POST | JSON (input text, model, voice, speed) | WAV audio bytes | OpenAI-compatible |
| `/v1/audio/process` | POST | Multipart audio file + model | WAV audio bytes | oMLX-specific |

### Model detection

Auto-detection via `config.json` fields, using both `architectures` (priority) and `model_type` (fallback):

- **STT**: WhisperForConditionalGeneration, Qwen3ASRForConditionalGeneration, ParakeetForCTC; model_type ∈ {whisper, qwen3_asr, parakeet}
- **TTS**: KokoroForConditionalGeneration, Qwen3TTSForConditionalGeneration, ChatterboxForConditionalGeneration; model_type ∈ {qwen3_tts, kokoro, chatterbox}
- **STS**: DeepFilterNetModel, MossFormer2SEModel, SAMAudio, LFM2AudioModel; model_type ∈ {deepfilternet, mossformer2, sam_audio}; or "lfm" prefix

Previously, `whisper` and `qwen3_tts` were in `UNSUPPORTED_MODEL_TYPES` and silently skipped during discovery. They are now properly detected and served.

### Dependency

```toml
[project.optional-dependencies]
audio = ["mlx-audio>=0.4.0"]
```

Install: `pip install omlx[audio]`

All mlx-audio imports are lazy (inside `start()` / `process()` methods). When not installed, LLM/VLM/Embedding/Reranker functionality is completely unaffected.

### Unit tests (108 passed, all mocked — no mlx-audio required)

```
tests/test_audio_stt.py         22 passed
tests/test_audio_tts.py         19 passed
tests/test_audio_sts.py         25 passed (4 integration deselected)
tests/test_audio_discovery.py   41 passed
tests/test_audio_api.py          7 passed (1 deselected)
tests/test_audio_memory.py       6 passed
────────────────────────────────────────
Total                          108 passed, 0 failed
```

### Integration tests (real models on Apple Silicon, mlx-audio 0.4.1)

**STT** — Verified with Whisper models (transcribe audio → text with segments and language detection).

**TTS** — Verified with Kokoro/Qwen3-TTS models (text → WAV audio output).

**STS** — All 4 families tested with real models:

| Model | HF Repo | Result |
|-------|---------|--------|
| DeepFilterNet | `mlx-community/DeepFilterNet-mlx` | ✅ 48kHz WAV, speech enhancement |
| MossFormer2 | `starkdmi/MossFormer2-SE` | ✅ 48kHz WAV, speech enhancement |
| SAMAudio | `mlx-community/sam-audio-base-fp16` | ✅ 24kHz WAV, source separation |
| LFM2.5-Audio | `mlx-community/LFM2.5-Audio-1.5B-6bit` | ✅ 24kHz WAV, speech-to-speech |

### Full test suite regression

```
3017 passed, 27 failed (all pre-existing), 37 deselected
```

All 27 failures are pre-existing and unrelated to audio changes.

## Known Limitations

- **No batching** — Audio engines process one request at a time. Concurrent requests are serialized through the MLX executor.
- **No streaming** — TTS and STS output the complete WAV after processing. No chunked/streaming audio delivery.
- **No resampling** — Each model expects audio at its native sample rate (e.g., DeepFilterNet requires 48kHz). The engine does not resample automatically.
- **SAMAudio `descriptions` parameter** — defaults to `["speech"]` but is not yet exposed through the API endpoint.
- **LFM2.5-Audio generation parameters** — `max_new_tokens`, `temperature`, `audio_temperature` are available as engine kwargs but not yet exposed through the API endpoint.

## How to Test

```bash
# Install with audio support
pip install -e '.[audio]'

# Start server with audio models in model directory
omlx serve --model-dir /path/to/models/

# STT
curl -X POST http://localhost:8000/v1/audio/transcriptions \
  -F file=@speech.wav -F model=whisper-large-v3

# TTS
curl -X POST http://localhost:8000/v1/audio/speech \
  -H "Content-Type: application/json" \
  -d '{"model":"kokoro-tts","input":"Hello world"}' \
  --output output.wav

# STS (speech enhancement)
curl -X POST http://localhost:8000/v1/audio/process \
  -F file=@noisy_speech.wav -F model=DeepFilterNet-mlx
```

### CODE REVIEW

File: omlx/api/audio_routes.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    except Exception as exc:
+        raise HTTPException(status_code=500, detail=str(exc)) from exc
+
+    # Save uploaded file to a temp path so the engine can open it by path
+    suffix = os.path.splitext(file.filename or "audio.wav")[1] or ".wav"
+    tmp_path = None
+    try:
+        with tempfile.NamedTemporaryFile(delete=False, suffix=suffix) as tmp:
+            tmp_path = tmp.name
+            content = await file.read()
```
Comment (anthrotype) [inline omlx/api/audio_routes.py]: readoing the entire upload into memory with no limit could exhaust server memory e.g. multi-GB upload. Perhaps you could add a MAX_AUDIO_SIZE check (e.g., 100MB?) before reading the full content

File: omlx/api/audio_routes.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    temperature: float = Form(0.0),
+):
+    """OpenAI-compatible audio transcription endpoint (Speech-to-Text)."""
+    from omlx.exceptions import ModelNotFoundError
+
+    pool = _get_engine_pool()
+
+    # Load the engine via pool (handles model loading and LRU eviction)
+    try:
+        engine = await pool.get_engine(model)
```
[the remark this answers is not in the document]
Comment (ethannortharc) [reply omlx/api/audio_routes.py]: Agreed — an AttributeError would be confusing. Added isinstance checks for STTEngine, TTSEngine, and STSEngine in all three endpoints. Wrong model type now returns a clear 400 error like "Model 'x' is not a speech-to-text model". Fixed in 23413c

**Comment by ethannortharc** [comment]
> another thing I forgot to mention, right now there's an undeclared dependency on `python-multipart`. This should be added to the pyproject.toml main dependencies, rather than in [audio] optional dependencies because the audio routes are registered unconditionally at omlx/server.py. This import happens at server startup regardless of whether mlx-audio is installed. And audio_routes.py uses File() and Form() from FastAPI, which require python-multipart at route registration time. Alternatively you could make the audio router registration lazy/conditional and only import and register it when mlx-audio is installed.

Good catch on the undeclared dependency. Went with the conditional registration approach — the audio router is now
wrapped in a try/except ImportError in server.py, so oMLX works cleanly without python-multipart when audio support
isn't needed. This keeps python-multipart out of the core dependencies and aligns with the design where mlx-audio (and its transitive deps) are fully optional

**Comment by ethannortharc** [comment]
Thanks for the thorough review, @anthrotype! All seven points have been addressed in commit 23413c1. Really
  appreciate the detailed feedback — especially the strict=False concern, that was an important catch. Let me know if
  anything else needs attention.

**Comment by anthrotype** [comment]
python-multipart should still be added to `[audio]` optional deps. Also, there currently is no request-level logging in TTS/STT engines, only start/stop is logged, which makes it hard to debug issues or monitor usage.

**Comment by ethannortharc** [comment]
> python-multipart should still be added to `[audio]` optional deps. Also, there currently is no request-level logging in TTS/STT engines, only start/stop is logged, which makes it hard to debug issues or monitor usage.

 @anthrotype  Good points, both addressed in a5a174b:

 > python-multipart: Added python-multipart>=0.0.5 to the [audio] optional dependencies. Combined with the
 conditional router registration from the previous fix, this means: pip install omlx[audio] pulls in both mlx-audio
 and python-multipart, while a plain pip install omlx skips both and the audio routes aren't registered.

 > Request-level logging: Added logger.info calls at the start and end of each request in all three engines:
 > - STT: logs model name, filename, file size, language at start; elapsed time and output text length at completion
 > - TTS: logs model name, text length, voice, speed at start; elapsed time and output size at completion
 > - STS: logs model name, family, filename, file size at start; elapsed time and output size at completion

 Thanks again for the follow-up feedback!

**Comment by iazrael** [comment]
非常期待这个特性！

**Comment by ynqjwsm** [comment]
really need this ❤️

**Comment by jundot** [comment]
@ethannortharc sorry for the slow response. As i mentioned in the discussions, my day job + critical bug fixes eat up most of my weekdays, so PRs tend to pile up. I finally got to sit down with this one and i'm glad i did.

This is seriously impressive work. I had audio support (mlx-audio integration) on my roadmap and was about to start on it, so seeing this PR show up was a really nice surprise. The way you integrated into the existing architecture (BaseNonStreamingEngine, EnginePool LRU, model discovery) is exactly how i would've approached it. The dynamic detection from mlx-audio's MODEL_REMAPPING is a great idea too. And 108 unit tests on top of that. Really well done.

I went through the code carefully and found a few things that need fixing before i can merge. Nothing major, mostly edge cases.

### needs fixing before merge

**STS family detection can crash at runtime**

This is the main one. `model_discovery.py` detects audio_sts models by `config.json` architectures/model_type, but `sts.py`'s `_detect_sts_family()` detects the family by model name string matching. These two can disagree.

For example, a model with `"architectures": ["DeepFilterNetModel"]` but directory name `my-custom-enhancer` gets discovered as audio_sts just fine, but then `_detect_sts_family("my-custom-enhancer")` returns `"generic"` and `start()` raises `ValueError`. So any model whose directory name doesn't contain "deepfilter"/"mossformer"/"sam"+"audio"/"lfm" will crash on load even though discovery says it's a valid STS model.

I think the cleanest fix is to pass the family info from config.json (architectures or model_type) through EngineEntry into the engine constructor, instead of re-detecting from the name string. Or at least fall back to config.json when name matching returns "generic".

**`_detect_sts_family` string matching is too broad**

`"sam" in name_lower and "audio" in name_lower` is just a 3-letter substring check, and `"lfm" in name_lower` is the same story. I'm worried a regular llm/vlm model could get misdetected as STS with matching this loose. Could you tighten these up? Maybe check for word boundaries or use more specific patterns like `sam-audio` or `sam_audio`.

**`_read_upload()` loads entire file before checking size**

Right now it does `content = await file.read()` first, then checks `len(content) > MAX_AUDIO_UPLOAD_BYTES`. So a 1GB upload already sits in memory before getting rejected. Would be better to read in chunks and bail early if it exceeds the limit.

**`server.py` try/except ImportError is dead code**

```python
try:
    from .api.audio_routes import router as audio_router
    app.include_router(audio_router)
except ImportError:
    pass
```

`audio_routes.py` only imports fastapi/stdlib stuff at module level, no mlx-audio. So this ImportError never fires and the routes are always registered regardless of whether mlx-audio is installed. The comment says "registered conditionally" but it's not actually conditional. Not a crash risk (the engines handle missing mlx-audio gracefully at start() time), but the code and comment are misleading. Either remove the try/except or add a real check.

### minor stuff (i'll clean these up after merge)

- `AudioSegment` in `audio_models.py` is defined but never used (`AudioTranscriptionResponse.segments` is `List[dict]`)
- `_FAMILY_PROCESSORS` wraps functions in unnecessary lambdas, can just use direct function references
- `_build_audio_detection_sets()` runs at import time and scans mlx-audio's filesystem. If mlx-audio is installed but broken in a non-ImportError way, this could prevent the server from starting at all
- `import struct` in `test_audio_stt.py` and `test_audio_tts.py` is unused
- `_LLM_TYPE_COLLISIONS` needs manual maintenance when new LLM families collide with mlx-audio directory names
- `venvstacks.toml` needs mlx-audio added to the framework requirements for DMG builds. I'll handle this one myself after merge.

Thanks again for putting this together. Looking forward to getting this merged.

**Comment by ethannortharc** [comment]
@jundot Thank you so much for the thorough review and the kind words — really means a lot coming from the project
  author. All four items have been addressed in commit c6c1eed.

  ### 1. STS family detection can crash at runtime

  You were right that `_detect_sts_family()` re-guessing from the directory name was fragile and could disagree with
  `model_discovery.py`.

  Fixed by passing `config_model_type` from `EngineEntry` into the `STSEngine` constructor. Family detection now
  follows a three-tier priority:

  1. **config.json `model_type`** (via `EngineEntry.config_model_type`) — most reliable
  2. **config.json `architectures`** (read from disk if model_type didn't match) — catches custom directory names
  3. **Name-based heuristic** — fallback only, for HuggingFace repo names that haven't been downloaded yet

  A `_CONFIG_TYPE_TO_FAMILY` mapping table covers all known model_type and architecture values. A model with
  `"architectures": ["DeepFilterNetModel"]` in a directory named `my-custom-enhancer` now correctly resolves to
  `deepfilternet`.

  ### 2. `_detect_sts_family` string matching is too broad

  Tightened the name-based fallback patterns:
  - `"sam" + "audio"` → now requires `"sam-audio"` or `"sam_audio"` (a model named `samsung-audio-tool` won't match)
  - `"lfm"` → now requires `"lfm2"`, `"lfm-audio"`, or `"lfm_audio"` (a model named `platform-tools` won't match)

  Since config.json detection is now primary, the name-based path is rarely exercised anyway.

  ### 3. `_read_upload()` loads entire file before checking size

  Rewritten to read in 1 MB chunks and bail immediately when `MAX_AUDIO_UPLOAD_BYTES` is exceeded. A 1 GB upload now
  gets rejected after reading ~100 MB, not after loading the full file.

  ### 4. `server.py` try/except ImportError is dead code

  Good catch — `audio_routes.py` only imports fastapi/stdlib at module level, so the `ImportError` never fired.
  Replaced with an explicit `import mlx_audio` check. The audio routes are now genuinely only registered when
  mlx-audio is installed.

  ---

  Regarding the minor items — happy to fix any of those now if you'd prefer, or leave them for your post-merge cleanup
   as you mentioned. Just let me know.

  Thanks again for taking the time to review this carefully!

**Comment by jundot** [comment]
@ethannortharc v0.3.0rc1 is out with your audio integration included. https://github.com/jundot/omlx/releases/tag/v0.3.0rc1 — if you get a chance, please give it a test and let me know if anything looks off. thanks!

**Comment by ethannortharc** [comment]
> @ethannortharc v0.3.0rc1 is out with your audio integration included. https://github.com/jundot/omlx/releases/tag/v0.3.0rc1 — if you get a chance, please give it a test and let me know if anything looks off. thanks!

@jundot Thanks for the tag! I'll pull the RC and run through the audio integration paths this week. Will report back if anything looks off. 🎉

**Comment by ethannortharc** [comment]
@jundot Tried to verify but blocked by the dependency conflict in #462 — ``mlx-audio`` pins ``mlx-lm==0.31.1`` (latest release) while omlx requires the unreleased ``0.31.2 (564281f)``. Since that commit only adds benchmark delay
  support, reverting to ``v0.31.1`` might be the quickest fix. Happy to re-test once it's resolved.

**Comment by krmao** [comment]
Model type cosyvoice3 not supported. ??? @ethannortharc
